### ISSUE TITLE
readOnly support

### ISSUE THREAD

**Issue by jayvdb** [oxidecomputer/typify#348]
Typically use of `readOnly` results in two schemas, one for put/post/patch without the `readOnly` fields, and the other for get with all of the fields.

IMO it would be sufficient for typify to create two schemas with a predictable type name suffix for the smaller schema.

However it would be nice if somehow the two schema types can indicate that it has a pair, so macros built on top of typify types can programmatically know about the soft-link between them.

**Comment by ahl** [comment]
That's interesting. `schemars` equates `readOnly` with `serde`'s `skip_deserialize`... which makes sense if, say, you're a server publishing an OpenAPI description of types. I could imagine client generation with the opposite correlation: `readOnly` -> `skip_serialize`.

It would be interesting to generate types that excluded either `readOnly` or `writeOnly` fields. This could be either a generation option or submodules (e.g. `read` and `write`).

What's your use case?

**Comment by jayvdb** [comment]
We have a Rust service which emits JSON. Some of the fields can be written to and sent back to the service, but other fields in the same struct are not writable - one non-writable field is the record identifier, but there are other struct members which we don't support modification.  We're using https://github.com/juhaku/utoipa on progenitor emitted structs to build the API.

Currently we have two schema - the real one and the "writable" one - which is duplication we can remove/automate if we can start using `readOnly`.  We are using the generated structs on both the server and the client, so in our use-case it is simpler if there is a separate "writable" struct, with from/into between the complete and "writable" structs.  The downside to this is that if a struct with some readonly members is part of another struct, the containing struct also needs a "writable" struct, etc, etc.  In our case there is only two levels, so this is managable.

If typify was to use `skip_serialize` on the readonly fields, which makes sense for a client, we would need to remove that with `syn`, otherwise it would break our actix service code.  But we already have ~1000 lines of `syn` post-processing `progenitor`'s output, so a little more wont hurt.  Probably what we would do is use `syn` to detect however `typify` is representing the readonly-ness, and split the struct into two:
1) the server one using `serde`'s `skip_deserialize` to ensure that even if the incoming JSON has those members it will be stripped before it reaches the endpoint
2) the client one using `serde`'s `skip_serialize` so that it doesnt try to send that field.
And then some more `syn` so that the progenitor generated endpoint client code is using the appropriate struct.

In short, dont worry too much about our use-case.  As long as typify does *something* to represent readonly fields, we'll be happy and use `syn` to mold the generated code to meet our needs.
